lib/spack/spack/package.py

```diff
@@ -1673,39 +1673,62 @@ def all_patches(cls):
         return patches
 
     def content_hash(self, content=None):
-        """Create a hash based on the sources and logic used to build the
-        package. This includes the contents of all applied patches and the
-        contents of applicable functions in the package subclass."""
-        if not self.spec.concrete:
-            err_msg = ("Cannot invoke content_hash on a package"
-                       " if the associated spec is not concrete")
-            raise spack.error.SpackError(err_msg)
+        """Create a hash based on the artifacts and patches used to build this package.
 
-        hash_content = list()
-        try:
-            source_id = fs.for_package_version(self, self.version).source_id()
-        except fs.ExtrapolationError:
-            source_id = None
-        if not source_id:
-            # TODO? in cases where a digest or source_id isn't available,
-            # should this attempt to download the source and set one? This
-            # probably only happens for source repositories which are
-            # referenced by branch name rather than tag or commit ID.
-            env = spack.environment.active_environment()
-            from_local_sources = env and env.is_develop(self.spec)
-            if not self.spec.external and not from_local_sources:
-                message = 'Missing a source id for {s.name}@{s.version}'
-                tty.warn(message.format(s=self))
-            hash_content.append(''.encode('utf-8'))
-        else:
-            hash_content.append(source_id.encode('utf-8'))
-        hash_content.extend(':'.join((p.sha256, str(p.level))).encode('utf-8')
-                            for p in self.spec.patches)
+        This includes:
+            * source artifacts (tarballs, repositories) used to build;
+            * content hashes (``sha256``'s) of all patches applied by Spack; and
+            * canonicalized contents the ``package.py`` recipe used to build.
```

Comment:
This is a question, more than a comment. I assume:
1. Removing the "canonicalized" portion of `package.py` will simplify a bit how we need to compute the `package_hash` for the corresponding `package.py`, but will give more "false positives" - i.e. different hashes for things that have been effectively built the same (for instance if the only change is under a `@when` decorator that is not hit)
2. On the other hand even in the current state I think we have false positives (I don't think we introspect the code so much to say if the changes are in a branch of a conditional that is not hit, for instance)

Did we ever weigh the simplification in avoiding the canonicalization vs. the greater number of false positives we'd get?

Reply:
No, but I've run into this so it was a pain point for me -- I think if I modify a function *clearly* annotated `@when("@3.0")`, it's really annoying if it changes the hash for version `2.0`.  So I'm inclined to keep this optimization in, as we do know statically what can be removed.  I'm also trying to avoid churn in CI and reduce builds there, which I think is a big win, particularly for some of the more complex packages like ROCm, MFEM, LLVM, etc.  Look at the test examples in `lib/spack/spack/test/data/unparse` to see some of what I mean.

I generally think *anything* we can do to reduce false positives here is good.

If you want to play with this to make your own judgement, you can look at canonical source and hashes for packages with `spack pkg source --canonical <spec>` and `spack pkg hash <spec>`.